### DISCUSSION TITLE
On: workflow_run: does not work with release created

### DISCUSSION THREAD

**Discussion by geirolz** [community/community#26548] [Actions]
<p>Hi, I’m having some problems with GitHub Actions and I really can’t find any useful documentation about that.</p>
<p><strong>To summarize:</strong><br>
<em>Expected</em> = <strong>Create Release</strong>(UI) → <strong>CI</strong> runs → [once completed] <strong>CD</strong> runs<br>
<em>But I got</em> = <strong>Create Release</strong>(UI) → <strong>CI</strong> runs | <strong>CD</strong> is <strong>NOT</strong> triggered</p>
<p><strong>CI</strong></p>
<pre><code class="lang-auto">name: CI

# triggers
on:
  push:
    branches: ['*']
  release:
    types: [ created ]
  pull_request:
    branches:
      - master

# env variables
env:
  CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}

# jobs
jobs:
  build:

    runs-on: ubuntu-latest
    strategy:
      matrix:
        # supported scala versions
        scala:
          - 2.12.8
          - 2.13.4

    steps:
      - uses: actions/checkout@v2

      #----------- JDK -----------
      - name: Set up JDK 11
        uses: actions/setup-java@v1
        with:
          java-version: 11

      #----------- CACHE -----------
      - name: Cache SBT
        uses: actions/cache@v2.1.4
        with:
          # A list of files, directories, and wildcard patterns to cache and restore
          path: |
            ~/.ivy2/cache
            ~/.sbt
          # An explicit key for restoring and saving the cache
          key: ${{ runner.os }}-sbt-${{ matrix.scala }}-${{ hashFiles('**/build.sb') }}

      #----------- COMPILE -----------
      - name: Compile, Format, Test and Coverage for ${{ matrix.scala }}
        run: sbt ++${{ matrix.scala }} clean scalafmtCheck coverage test coverageReport
</code></pre>
<p><strong>CD</strong></p>
<pre><code class="lang-auto">name: CD

# triggers
on:
  workflow_run:
    workflows: ["CI"]
    branches:
      - master
    types:
      - completed

# env variables
env:
  SONATYPE_USERNAME: ${{ secrets.SONATYPE_USERNAME }}
  SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }}
  PGP_SECRET: ${{ secrets.PGP_SECRET }}

# jobs
jobs:
  deploy:
    runs-on: ubuntu-latest
    if: github.event_name != 'pull_request'
    steps:
      - uses: actions/checkout@v2

      #----------- JDK -----------
      - name: Set up JDK 11
        uses: actions/setup-java@v1
        with:
          java-version: 11

      #----------- CACHE -----------
      - name: Cache SBT
        uses: actions/cache@v2.1.4
        with:
          # A list of files, directories, and wildcard patterns to cache and restore
          path: |
            ~/.ivy2/cache
            ~/.sbt
          # An explicit key for restoring and saving the cache
          key: ${{ runner.os }}-sbt-${{ matrix.scala }}-${{ hashFiles('**/build.sb') }}

      #----------- RELEASE -----------
      - name: Release
        run: sbt clean compile ci-release
</code></pre>
<p>Can someone tell me why ? Am I wrong or is this a bug ?<br>
Thank you in advice for the help :slight_smile:</p>

**Comment by mickeygousset** [comment]
<p>Do you have both workflow files in the same repo?</p>

**Comment by geirolz** [comment]
<p>Yes for sure,  <a href="https://github.com/geirolz/advxml" rel="noopener nofollow ugc">this</a> is the project</p>
<p>CI runs, CD no<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/0/c/0cb3d8084915f97886e25386afccd3ff5b40ef37.png" data-download-href="/uploads/short-url/1On2GJ8EZN7c2iLZLuG12SpzBZB.png?dl=1" title="Screenshot 2021-04-28 at 16.35.56" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/31506564/181098424-6908190c-b538-4e7e-a4ed-501f7ba3ff71.png" alt="Screenshot 2021-04-28 at 16.35.56" data-base62-sha1="1On2GJ8EZN7c2iLZLuG12SpzBZB" width="690" height="64" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0cb3d8084915f97886e25386afccd3ff5b40ef37_2_690x64.png, https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0cb3d8084915f97886e25386afccd3ff5b40ef37_2_1035x96.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0cb3d8084915f97886e25386afccd3ff5b40ef37_2_1380x128.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/0/c/0cb3d8084915f97886e25386afccd3ff5b40ef37_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2021-04-28 at 16.35.56</span><span class="informations">1812×170 12.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by mickeygousset** [comment]
<p>huh. First glance it looks right, but we are missing something, logic or otherwise. I’ll take a look at it later today when I get a chance in more detail, and see if I can recreate it.</p>
<p>Question: What about your other triggers in the CI? For example, if you do a PR to master, not create a release, does that trigger the CD? Or does the CD still not run?</p>

**Comment by geirolz** [comment]
<p>Thank you!<br>
Yes CI works fine for PR on master and CD is triggered once CI is successfully completed, this creates the snapshot versions.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/e/4/e4528e721564e3ca7435bbdb488596a1ffd4a0d3.png" data-download-href="/uploads/short-url/wzPJR00JZt2z1JQ4UahjVmFhrur.png?dl=1" title="Screenshot 2021-04-28 at 16.41.12" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/31506564/181098425-6ca72a34-afcb-4c7a-92c5-aca2f955e0f6.png" alt="Screenshot 2021-04-28 at 16.41.12" data-base62-sha1="wzPJR00JZt2z1JQ4UahjVmFhrur" width="690" height="60" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/e/4/e4528e721564e3ca7435bbdb488596a1ffd4a0d3_2_690x60.png, https://aws1.discourse-cdn.com/github/optimized/3X/e/4/e4528e721564e3ca7435bbdb488596a1ffd4a0d3_2_1035x90.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/e/4/e4528e721564e3ca7435bbdb488596a1ffd4a0d3_2_1380x120.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/e/4/e4528e721564e3ca7435bbdb488596a1ffd4a0d3_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2021-04-28 at 16.41.12</span><span class="informations">1804×158 10.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by mickeygousset** [comment]
<p>interesting, so it only occurs when the Release creation triggers the CI.  Good to know.</p>

**Comment by geirolz** [comment]
<p>Exactly, is seems so. Once I create the Release CI works but once completed CD is not triggered</p>

**Comment by mickeygousset** [comment]
<p>Another question:  RIght now you have two workflow files CI and CD. Have you considered just doing it all in one workflow file instead?  Any reason you don’t want to do that?</p>

**Comment by geirolz** [comment]
<p>Just to keep them clean and separated, no other reasons honestly. Do you think this can be problematic?</p>

**Comment by mickeygousset** [comment]
<p>Not necessarily problematic.  But in general right now, putting things in one file can be a little simpler. You can still keep it clean by using multiple jobs.</p>

**Comment by geirolz** [comment]
<p>Ok thanks for the advice!</p>

**Comment by mickeygousset** [comment]
<p>That said I still want to figure out why this scenario isn’t working. Still looking into it.</p>

**Comment by geirolz** [comment]
<p>Any news? Did you replicate it?</p>

**Comment by mickeygousset** [comment, marked as answer]
<p>OK, I was able to recreate it, and I know why it is happening.</p>
<p>This repo is where I was working on it:  <a href="https://github.com/mickeygoussetorg/deleteme" class="inline-onebox">GitHub - mickeygoussetorg/deleteme</a></p>
<p>The problem is in your CD triggers, and in what is in the GitHub context object when you trigger off a created release vs when you just do a push.</p>
<p>If you just do a push, then the “ref” in the GitHub context object is refs/heads/main (or master, depending on what you use for your default branch). <a href="https://github.com/mickeygoussetorg/deleteme/runs/2497012353?check_suite_focus=true">You can see that here.</a></p>
<p>However, when you trigger off a release, the ref is instead: refs/tags/releasetag. <a href="https://github.com/mickeygoussetorg/deleteme/runs/2496996831?check_suite_focus=true">You can see that here</a>.</p>
<p>Therefor, when the CD does its check, one thing you have it checking for is branch: master.  Well, the CI wasn’t triggered from branch master it was triggered from refs/tags/releasetag, so you logic equates to false and it doesn’t run.</p>
<p>You can test this yourself by commenting out branch: master in your CD, and you’ll see that once you create  a release, the CI runs, and then the CD runs.</p>
<p>Its a logic issue. You need to rethink your logic.</p>

**Comment by geirolz** [comment]
<p>Great! Thank you!<br>
So maybe I can use this check instead <code>github.ref == 'refs/tags/releasetag'</code>, right ? I mean, if there is no possibility to check the release target branch.</p>
<p>UPDATE:<br>
Just tried but still not triggered</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/geirolz/advxml/actions/runs/810029954/workflow" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/31506564/181098432-2be5c3fd-4013-4dc2-af2f-6290ff1e45d2.jpeg" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/1/7/17bfeb087fd0ce2d01caf85e6eba1b9f8defb85a_2_690x345.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/1/7/17bfeb087fd0ce2d01caf85e6eba1b9f8defb85a_2_1035x517.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/3X/1/7/17bfeb087fd0ce2d01caf85e6eba1b9f8defb85a.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/1/7/17bfeb087fd0ce2d01caf85e6eba1b9f8defb85a_2_10x10.png"></div>

<h3><a href="https://github.com/geirolz/advxml/actions/runs/810029954/workflow" target="_blank" rel="noopener nofollow ugc">geirolz/advxml</a></h3>

<p><a href="https://github.com/geirolz/advxml/actions/runs/810029954/workflow" target="_blank" rel="noopener nofollow ugc">ns/runs/810029954/workflow</a></p>

  <p><span class="label1">A lightweight, simple and functional library DSL to work with XML in Scala with Cats</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by mickeygousset** [comment]
<p><code>github.ref == 'refs/tags/releasetag'</code></p>
<p>This line isn’t going to work, because you didn’t call your release “releasetag”.  You called it v2.3.1. So the logic is going to evaluate to false, because the tag didn’t match.</p>
<p>I’m out of office for the next few days, but I’ll see if I can wrap my head around a way to figure out the logic on this.</p>

**Comment by geirolz** [comment]
<p>I think we can close this ticket, I’ve resolved removing this check, it’s not mandatory actually.</p>
<p>Thank you so much <a class="mention" href="/mickeygousset">@mickeygousset</a> for your help!</p>
